### PR TITLE
gh-66143: Allow copying of CodecInfo object

### PR DESCRIPTION
**Pull request by furkanonder**
* Issue: gh-66143

### CODE REVIEW

File: Lib/test/test_codecs.py [hunk 1]
```diff
@@ -1762,6 +1762,58 @@ def test_file_closes_if_lookup_error_raised(self):
...
+        # Test a CodecInfo with _is_text_encoding equal to false.
+        orig = codecs.lookup("rot13")
+        dup = copy.copy(orig)
+        self.assertIsNot(dup, orig)
+        self.assertFalse(orig._is_text_encoding)
+        self.assertEqual(dup.encode, orig.encode)
+        self.assertEqual(dup.name, orig.name)
+        self.assertEqual(dup.incrementalencoder, orig.incrementalencoder)
+
+    def test_deep_copy(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_codecs.py]:
```suggestion
    def test_deepcopy(self):
```

File: Lib/test/test_codecs.py [hunk 2]
```diff
@@ -1762,6 +1762,58 @@ def test_file_closes_if_lookup_error_raised(self):
 
             file().close.assert_called()
 
+    def test_copy(self):
+        orig = codecs.lookup('utf-8')
+        dup = copy.copy(orig)
+        self.assertIsNot(dup, orig)
```
Comment (serhiy-storchaka) [inline Lib/test/test_codecs.py:1768]: Add also the equality test. Here and in other test methods.
```suggestion
        self.assertIsNot(dup, orig)
        self.assertEqual(dup, orig)
```

File: Lib/test/test_codecs.py [hunk 3]
```diff
@@ -1762,6 +1762,58 @@ def test_file_closes_if_lookup_error_raised(self):
...
+        self.assertEqual(dup.encode, orig.encode)
+        self.assertEqual(dup.name, orig.name)
+        self.assertEqual(dup.incrementalencoder, orig.incrementalencoder)
+
+        # Test a CodecInfo with _is_text_encoding equal to false.
+        orig = codecs.lookup("rot13")
+        dup = copy.deepcopy(orig)
+        self.assertIsNot(dup, orig)
+        self.assertFalse(orig._is_text_encoding)
+        self.assertNotEqual(dup.encode, orig.encode)
```
Comment (serhiy-storchaka) [inline Lib/test/test_codecs.py]: It is a red herring!

I think that it is better to use other codec as an example for `_is_text_encoding=False` in the tests. `rot_13` is an outlier.

File: Lib/test/test_codecs.py [hunk 4]
```diff
@@ -1762,6 +1762,58 @@ def test_file_closes_if_lookup_error_raised(self):
...
+        self.assertEqual(dup.name, orig.name)
+        self.assertEqual(dup.incrementalencoder, orig.incrementalencoder)
+
+    def test_pickle(self):
+        codec_info = codecs.lookup('utf-8')
+        for proto in range(pickle.HIGHEST_PROTOCOL + 1):
+            with self.subTest(protocol=proto):
+                pickled_codec_info = pickle.dumps(codec_info)
+                unpickled_codec_info = pickle.loads(pickled_codec_info)
+                self.assertEqual(codec_info, unpickled_codec_info)
```
Comment (serhiy-storchaka) [inline Lib/test/test_codecs.py:1812]: The equality test only tests encode, decode, streamreader, and streamwriter values. Add also tests for name, incrementalencoder and _is_text_encoding.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by vstinner** [comment]
> CodecInfo objects as retrieved from codecs.lookup currently throw an exception when trying to copy or pickle them.

How is it an issue? What is the use case?

**Pull request merged**